velox/functions/sparksql/Comparisons.cpp

```diff
@@ -0,0 +1,155 @@
+ */
+#include "velox/functions/sparksql/LeastGreatest.h"
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/functions/sparksql/Comparisons.h"
+#include "velox/type/Type.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+template <typename Cmp, TypeKind kind>
+class ComparisonFunction final : public exec::VectorFunction {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  bool isDefaultNullBehavior() const override {
+    return true;
+  }
+
+  bool supportsFlatNoNullsFastPath() const override {
+    return true;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, BOOLEAN(), result);
+    auto* flatResult = result->asFlatVector<bool>();
+    flatResult->mutableRawValues<uint64_t>();
+    const Cmp cmp;
+    if (args[0]->isFlatEncoding() && args[1]->isFlatEncoding()) {
+      // Fast path for (flat, flat).
+      auto flatA = args[0]->asUnchecked<FlatVector<T>>();
+      auto rawA = flatA->mutableRawValues();
+      auto flatB = args[1]->asUnchecked<FlatVector<T>>();
+      auto rawB = flatB->mutableRawValues();
+      rows.applyToSelected(
+          [&](vector_size_t i) { flatResult->set(i, cmp(rawA[i], rawB[i])); });
+    } else if (args[0]->isConstantEncoding() && args[1]->isFlatEncoding()) {
+      // Fast path for (const, flat).
+      auto constant = args[0]->asUnchecked<SimpleVector<T>>()->valueAt(0);
+      auto flatValues = args[1]->asUnchecked<FlatVector<T>>();
+      auto rawValues = flatValues->mutableRawValues();
+      rows.applyToSelected([&](vector_size_t i) {
+        flatResult->set(i, cmp(constant, rawValues[i]));
+      });
+    } else if (args[0]->isFlatEncoding() && args[1]->isConstantEncoding()) {
+      // Fast path for (flat, const).
+      auto flatValues = args[0]->asUnchecked<FlatVector<T>>();
+      auto constant = args[1]->asUnchecked<SimpleVector<T>>()->valueAt(0);
+      auto rawValues = flatValues->mutableRawValues();
+      rows.applyToSelected([&](vector_size_t i) {
+        flatResult->set(i, cmp(rawValues[i], constant));
+      });
+    } else {
+      // Fast path if one or more arguments are encoded.
+      exec::DecodedArgs decodedArgs(rows, args, context);
+      auto decoded0 = decodedArgs.at(0);
+      auto decoded1 = decodedArgs.at(1);
+      rows.applyToSelected([&](vector_size_t i) {
+        flatResult->set(
+            i, cmp(decoded0->valueAt<T>(i), decoded1->valueAt<T>(i)));
+      });
+    }
+  }
+};
+
+template <template <typename> class Cmp>
+std::shared_ptr<exec::VectorFunction> makeImpl(
+    const std::string& functionName,
+    const std::vector<exec::VectorFunctionArg>& args) {
+  VELOX_CHECK_EQ(args.size(), 2);
+  for (size_t i = 1; i < args.size(); i++) {
+    VELOX_CHECK(*args[i].type == *args[0].type);
```

Comment:
How do we support complex type. The function comment confused me since they are conflicted.
```
  /// Types are weakly matched.
  /// Examples: Two RowTypes are equivalent if the children types are
  /// equivalent, but the children names could be different. Two OpaqueTypes are
  /// equivalent if the typeKind matches, but the typeIndex could be different.
  virtual bool equivalent(const Type& other) const = 0;

  /// Types are strongly matched.
  /// Examples: Two RowTypes are == if the children types and the children names
  /// are same. Two OpaqueTypes are == if the typeKind and the typeIndex are
  /// same. Same as equivalent for most types except for Row, Opaque types.
  virtual bool operator==(const Type& other) const {
    return this->equivalent(other);
  }
```

Reply:
This is referred from `LeastGreatest.cpp`. When will it happen that the name is not same?

Reply:
`Two RowTypes are == if the children types and the children names
  /// are same. `

Reply:
Functions here are only for scalar types so it should not be `RowTypes` provided here but should be the leaf child type.